---
title: "Class 11: Structural Bioinformatics pt 2"
author: "Montserrat (PID:A16536527)"
format: pdf
toc: TRUE
---

## Background 

we saw last day that the PDB has 209.886 entries (Oct/Nov 20235). UniProtKB (i.e. protein sequence database) has 199,579,901 entries.

```{r}
209886/199579901 * 100
```

so the PDB has only 0.1% coverage of the main sequence database.

Enter AlphaFold data base (AFDB) < https://alphafold.ebi.ac.uk > taht attempts to provide computed models for all sequences in UniProt. 

accoridng to website, AlphaFold DB provides open access to over 200 million protein structure predictions to accelerate scientific research.


##AlphaFold 

AlphaFold has 3 main outpouts 
- predicted coordinates (PDB files)
-local quality score called pLDDD (one for each amino-acid)
- a second quality score PAE Predicted Aligned Error (for each pair of amino-acid)

We can run AlphaFold ourselves if we are not happy with AFDB (i.e. no coverage or poor model)

## Interpreting/ analyzing AF results in R 

```{r}
results_dir <- "HIVPR_dimer_23119/" 
```


```{r}
pdb_files <- list.files(path=results_dir,
                        pattern="*.pdb",
                        full.names = TRUE)
# Print our PDB file names
basename(pdb_files)
```


```{r}
library(bio3d)

# Read all data from Models 
#  and superpose/fit coords
pdbs <- pdbaln(pdb_files, fit=TRUE, exefile="msa")
```


```{r}
pdbs
```
RMSD is a standard measure of structural distance between coordinate sets. We can use the rmsd() function to calculate the RMSD between all pairs models.

```{r}
rd <- rmsd(pdbs, fit=T)

range(rd)
```
Draw a heatmap of these RMSD matrix values

```{r}
library(pheatmap)

colnames(rd) <- paste0("m",1:5)
rownames(rd) <- paste0("m",1:5)
pheatmap(rd)
```

## Alignment File 


```{r}
aln_file <- list.files(path=results_dir,
                       pattern=".a3m$",
                        full.names = TRUE)
aln_file
```

